x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/oidc/TransportOpenIdConnectLogoutAction.java

```diff
@@ -0,0 +1,142 @@
+            (Writeable.Reader<OpenIdConnectLogoutRequest>) OpenIdConnectLogoutRequest::new);
+        this.realms = realms;
+        this.tokenService = tokenService;
+    }
+
+    @Override
+    protected void doExecute(Task task, OpenIdConnectLogoutRequest request, ActionListener<OpenIdConnectLogoutResponse> listener) {
+        invalidateRefreshToken(request.getRefreshToken(), ActionListener.wrap(ignore -> {
+            try {
+                final String token = request.getToken();
+                tokenService.getAuthenticationAndMetaData(token, ActionListener.wrap(
+                    tuple -> {
+                        final Authentication authentication = tuple.v1();
+                        final Map<String, Object> tokenMetadata = tuple.v2();
+                        validateAuthenticationAndMetadata(authentication, tokenMetadata);
+                        tokenService.invalidateAccessToken(token, ActionListener.wrap(
+                            result -> {
+                                if (logger.isTraceEnabled()) {
+                                    logger.trace("OpenID Connect Logout for user [{}] and token [{}...{}]",
+                                        authentication.getUser().principal(),
+                                        token.substring(0, 8),
+                                        token.substring(token.length() - 8));
+                                }
+                                OpenIdConnectLogoutResponse response = buildResponse(authentication, tokenMetadata);
+                                listener.onResponse(response);
+                            }, listener::onFailure)
+                        );
+                    }, listener::onFailure));
+            } catch (IOException e) {
+                logger.debug("Internal error during OpenID Connect Logout");
+                listener.onFailure(e);
+            }
+        }, listener::onFailure));
+    }
+
+    private OpenIdConnectLogoutResponse buildResponse(Authentication authentication, Map<String, Object> tokenMetadata) {
+        final String idTokenHint = (String) getFromMetadata(tokenMetadata, "id_token_hint");
+        final Realm realm = this.realms.realm(authentication.getAuthenticatedBy().getName());
+        final JWT idToken;
+        try {
+            idToken = JWTParser.parse(idTokenHint);
+        } catch (ParseException e) {
+            throw new ElasticsearchSecurityException("Token Metadata did not contain a valid IdToken", e);
+        }
+        return ((OpenIdConnectRealm) realm).buildLogoutResponse(idToken);
+    }
+
+    private void validateAuthenticationAndMetadata(Authentication authentication, Map<String, Object> tokenMetadata) {
+        if (tokenMetadata == null) {
+            throw new ElasticsearchSecurityException("Authentication did not contain metadata");
+        }
+        if (authentication == null) {
+            throw new ElasticsearchSecurityException("No active authentication");
```

Comment:
nit: I see these map to HTTP 500, I think 400 is better. So maybe throw `IllegalArgumentException`. I don't have a consistent perspective over these stuff (across all APIs), but 500 is the server's fault and 400 is the client's .

Reply:
This is the client giving the server a valid access token and the server fetching a UserToken from .security that has no `Authentication`. I would claim this is the server's error if the UserToken was malformed in storage or if it originally created one with no Authentication

Reply:
You're correct, I was putting all the exceptions thrown in this method in the same bucket. This one is indeed internal server error.

Reply:
I'll change the following: 
```
if (realm instanceof OpenIdConnectRealm == false) {
            throw new IllegalArgumentException("Access token is not valid for an OpenID Connect realm");
}
```
This can't happen with Kibana, but we mean to allow these APIs to be used by custom web apps that might support more than one authentication realms and in that case these apps should use the correct rest API for the given realm that authenticated their user. Thanks for the feedback on this one